src/DefaultBuilder/src/ConfigureWebHostBuilder.cs

```diff
@@ -46,7 +46,8 @@ public IWebHostBuilder ConfigureAppConfiguration(Action<WebHostBuilderContext, I
         // Run these immediately so that they are observable by the imperative code
         configureDelegate(_context, _configuration);
 
-        if (!string.Equals(HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot), HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.WebRootKey], previousContentRoot), StringComparison.OrdinalIgnoreCase))
+        if (_configuration[WebHostDefaults.WebRootKey] != null &&
+                !string.Equals(HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot), HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.WebRootKey], previousContentRoot), StringComparison.OrdinalIgnoreCase))
```

Comment:
```suggestion
        if (_configuration[WebHostDefaults.WebRootKey] is {} webRootKey &&
            !string.Equals(HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot), HostingPathResolver.ResolvePath(webRootKey, previousContentRoot), StringComparison.OrdinalIgnoreCase))
```

Reply:
The issue is this is throwing when the configuration isn't actually changing. There's some special logic in this type to ensure that environmental settings that cannot be modified like web root and content root.

Basically, this makes `ConfigureWebHostBuilder.UseWebRoot(path)` and `ConfigureWebHostBuilder.UseContentRoot(path)` throw only if they actually change the directory rather than reset the value. This logic is trying to be smart and detect if the new path was really the old path expressed differently, but can fire in some cases when the config hasn't changed at all in this method.

We should make the following change. We should **never** throw for config that hasn't changed. It is possible for `_configuration[WebHostDefaults.WebRootKey]` to diverge from the real `_context.HostingEnvironment.WebRootPath` because these guards aren't on people accessing this `_configuration` via `WebApplicationBuilder.Configuration`.

I'm surprised more people haven't run into this. We may want to service it. Unfortunately, I'm leaving on vacation tonight. @Tratcher can you handle this?

```diff
diff --git a/src/DefaultBuilder/src/ConfigureHostBuilder.cs b/src/DefaultBuilder/src/ConfigureHostBuilder.cs
index 9e388a4dad..285cc94fe8 100644
--- a/src/DefaultBuilder/src/ConfigureHostBuilder.cs
+++ b/src/DefaultBuilder/src/ConfigureHostBuilder.cs
@@ -62,6 +62,7 @@ public sealed class ConfigureHostBuilder : IHostBuilder, ISupportsConfigureWebHo
         var previousApplicationName = _configuration[HostDefaults.ApplicationKey];
         // Use the real content root so we can compare paths
         var previousContentRoot = _context.HostingEnvironment.ContentRootPath;
+        var previousContentRootConfig = _configuration[HostDefaults.ContentRootKey];
         var previousEnvironment = _configuration[HostDefaults.EnvironmentKey];

         // Run these immediately so that they are observable by the imperative code
@@ -74,7 +75,8 @@ public sealed class ConfigureHostBuilder : IHostBuilder, ISupportsConfigureWebHo
             throw new NotSupportedException($"The application name changed from \"{previousApplicationName}\" to \"{_configuration[HostDefaults.ApplicationKey]}\". Changing the host configuration using WebApplicationBuilder.Host is not supported. Use WebApplication.CreateBuilder(WebApplicationOptions) instead.");
         }

-        if (!string.Equals(previousContentRoot, HostingPathResolver.ResolvePath(_configuration[HostDefaults.ContentRootKey]), StringComparison.OrdinalIgnoreCase))
+        if (!string.Equals(previousContentRootConfig, _configuration[HostDefaults.ContentRootKey], StringComparison.OrdinalIgnoreCase)
+            && !string.Equals(previousContentRoot, HostingPathResolver.ResolvePath(_configuration[HostDefaults.ContentRootKey]), StringComparison.OrdinalIgnoreCase))
         {
             throw new NotSupportedException($"The content root changed from \"{previousContentRoot}\" to \"{HostingPathResolver.ResolvePath(_configuration[HostDefaults.ContentRootKey])}\". Changing the host configuration using WebApplicationBuilder.Host is not supported. Use WebApplication.CreateBuilder(WebApplicationOptions) instead.");
         }
diff --git a/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs b/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs
index e0e125657e..ab0feebeb2 100644
--- a/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs
+++ b/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs
@@ -37,7 +37,9 @@ public sealed class ConfigureWebHostBuilder : IWebHostBuilder, ISupportsStartup
     public IWebHostBuilder ConfigureAppConfiguration(Action<WebHostBuilderContext, IConfigurationBuilder> configureDelegate)
     {
         var previousContentRoot = _context.HostingEnvironment.ContentRootPath;
+        var previousContentRootConfig = _configuration[WebHostDefaults.ContentRootKey];
         var previousWebRoot = _context.HostingEnvironment.WebRootPath;
+        var previousWebRootConfig = _configuration[WebHostDefaults.WebRootKey];
         var previousApplication = _configuration[WebHostDefaults.ApplicationKey];
         var previousEnvironment = _configuration[WebHostDefaults.EnvironmentKey];
         var previousHostingStartupAssemblies = _configuration[WebHostDefaults.HostingStartupAssembliesKey];
@@ -46,7 +48,8 @@ public sealed class ConfigureWebHostBuilder : IWebHostBuilder, ISupportsStartup
         // Run these immediately so that they are observable by the imperative code
         configureDelegate(_context, _configuration);

-        if (!string.Equals(HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot), HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.WebRootKey], previousContentRoot), StringComparison.OrdinalIgnoreCase))
+        if (!string.Equals(previousWebRootConfig, _configuration[WebHostDefaults.WebRootKey], StringComparison.OrdinalIgnoreCase)
+            && !string.Equals(HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot), HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.WebRootKey], previousContentRoot), StringComparison.OrdinalIgnoreCase))
         {
             // Diasllow changing the web root for consistency with other types.
             throw new NotSupportedException($"The web root changed from \"{HostingPathResolver.ResolvePath(previousWebRoot, previousContentRoot)}\" to \"{HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.WebRootKey], previousContentRoot)}\". Changing the host configuration using WebApplicationBuilder.WebHost is not supported. Use WebApplication.CreateBuilder(WebApplicationOptions) instead.");
@@ -56,7 +59,8 @@ public sealed class ConfigureWebHostBuilder : IWebHostBuilder, ISupportsStartup
             // Disallow changing any host configuration
             throw new NotSupportedException($"The application name changed from \"{previousApplication}\" to \"{_configuration[WebHostDefaults.ApplicationKey]}\". Changing the host configuration using WebApplicationBuilder.WebHost is not supported. Use WebApplication.CreateBuilder(WebApplicationOptions) instead.");
         }
-        else if (!string.Equals(previousContentRoot, HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.ContentRootKey]), StringComparison.OrdinalIgnoreCase))
+        else if (!string.Equals(previousContentRootConfig, _configuration[WebHostDefaults.ContentRootKey], StringComparison.OrdinalIgnoreCase)
+            && !string.Equals(previousContentRoot, HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.ContentRootKey]), StringComparison.OrdinalIgnoreCase))
         {
             // Disallow changing any host configuration
             throw new NotSupportedException($"The content root changed from \"{previousContentRoot}\" to \"{HostingPathResolver.ResolvePath(_configuration[WebHostDefaults.ContentRootKey])}\". Changing the host configuration using WebApplicationBuilder.WebHost is not supported. Use WebApplication.CreateBuilder(WebApplicationOptions) instead.");
```

In .NET 7, I suspect we'll stop throwing from here when the content root changes, because the never actually modified the content root when applied via `ConfigureAppConfiguration`, but that's not a pressing concern.

Reply:
@[Tratcher](https://github.com/Tratcher)  have changed, please review